File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/infra/DefaultWebSocketUnifiedService.java
package com.kucoin.universal.sdk.internal.infra;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.internal.interfaces.WebSocketUnifiedService;
import com.kucoin.universal.sdk.internal.interfaces.WebsocketTransport;
import com.kucoin.universal.sdk.internal.interfaces.WebsocketTransportListener;
import com.kucoin.universal.sdk.model.*;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class DefaultWebSocketUnifiedService
    implements WebSocketUnifiedService, WebsocketTransportListener<UnifiedWsMessage> {

  private final WebSocketClientOption option;
  private final WebsocketTransport client;
  private final ObjectMapper mapper = new ObjectMapper();
  private final Map<String, CompletableFuture<UnifiedWsMessage>> futureMap =
      new ConcurrentHashMap<>();

  public DefaultWebSocketUnifiedService(ClientOption opt, String sdkVersion) {
    if (opt.getUnifiedWsEndpoint() == null || opt.getUnifiedWsEndpoint().isEmpty()) {
      throw new IllegalArgumentException("Missing unified WebSocket endpoint");
    }

    mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
    this.option = opt.getWebsocketClientOption();
    this.client =
        new DefaultWebsocketTransport(
            new DefaultWebsocketUnifiedMetaProvider(opt, mapper), option, this);
  }

  @Override
  public void onEvent(WebSocketEvent event, String message) {
    notifyEvent(event, message);
  }

  @Override
  public void onMessage(UnifiedWsMessage wsMessage) {
    CompletableFuture<UnifiedWsMessage> future = futureMap.remove(wsMessage.getId());
    if (future != null) {
      if (future.complete(wsMessage)) {
        log.debug(
            "Unified WebSocket API response received, id={}, op={}",
            wsMessage.getId(),
            wsMessage.getOp());
      } else {
        log.warn("Future for id={} was already completed", wsMessage.getId());
      }
    }
  }

  @Override
  public void onReconnected() {
    // No recovery needed for unified API; subscriptions are managed elsewhere
  }

  @Override
  public void start() {
    client.start();
  }

  @Override
  public void stop() {
    client.stop();
  }

  @Override
  public CompletableFuture<UnifiedWsMessage> call(UnifiedWsArgs args) {
    // Ensure ID exists
    if (args.getId() == null || args.getId().isEmpty()) {
      args.setId(UUID.randomUUID().toString());
    }

    // Ensure operation type is set
    if (args.getOp() == null || args.getOp().isEmpty()) {
      throw new IllegalArgumentException("Missing required field: op");
    }

    CompletableFuture<UnifiedWsMessage> future = new CompletableFuture<>();

    // Check for duplicate ID
    Object previous = futureMap.put(args.getId(), future);
    if (previous != null) {
      throw new IllegalArgumentException("Duplicate request id: " + args.getId());
    }

    log.info("Calling unified WebSocket API, id={}, op={}", args.getId(), args.getOp());

    return client
        .write(args.getId(), args, option.getWriteTimeout())
        .thenApply(
            ignored -> {
              log.info("Unified WebSocket API call sent, id={}, op={}", args.getId(), args.getOp());
              return future;
            })
        .exceptionally(
            ex -> {
              futureMap.remove(args.getId());
              log.error(
                  "Unified WebSocket API call failed, id={}, op={}",
                  args.getId(),
                  args.getOp(),
                  ex);
              future.completeExceptionally(ex);
              return future;
            })
        .thenCompose(Function.identity());
  }

  private void notifyEvent(WebSocketEvent ev, String msg) {
    if (option.getEventCallback() != null) {
      try {
        option.getEventCallback().onEvent(ev, msg);
      } catch (Exception e) {
        log.error("exception when notify event", e);
      }
    }
  }
}

File: sdk/php/src/Internal/Infra/DefaultWebSocketUnifiedService.php
<?php

namespace KuCoin\UniversalSDK\Internal\Infra;

use Exception;
use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketClient;
use KuCoin\UniversalSDK\Internal\Interfaces\WebSocketUnifiedService;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOption;
use KuCoin\UniversalSDK\Model\UnifiedWsArgs;
use KuCoin\UniversalSDK\Model\UnifiedWsMessage;
use React\EventLoop\LoopInterface;
use React\Promise\Deferred;
use React\Promise\PromiseInterface;
use RuntimeException;

/**
 * Default implementation of the WebSocketUnifiedService interface.
 */
class DefaultWebSocketUnifiedService implements WebSocketUnifiedService
{
    /**
     * @var ClientOption
     */
    private $option;

    /**
     * @var WebSocketClient
     */
    private $client;

    /**
     * @var Serializer
     */
    private $serializer;

    /**
     * Map of request IDs to their corresponding promises
     * @var array<string, Deferred>
     */
    private $futureMap = [];

    /**
     * @param ClientOption $opt
     * @param LoopInterface $loop
     * @param string $sdkVersion
     */
    public function __construct(ClientOption $opt, LoopInterface $loop, string $sdkVersion)
    {
        if (empty($opt->unifiedWsEndpoint)) {
            throw new RuntimeException("Missing unified WebSocket endpoint");
        }

        $this->option = $opt->websocketClientOption;
        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(new JsonSerializedHandler());
            })->build();

        $this->client = new DefaultWebSocketClient(
            new DefaultWebsocketUnifiedMetaProvider($opt, $this->serializer),
            $this->option,
            $loop
        );

        $this->client->on("event", function ($event, $msg) {
            $this->notifyEvent($event, $msg);
        });

        $this->client->on("message", function (UnifiedWsMessage $wsMessage) {
            $deferred = $this->futureMap[$wsMessage->id] ?? null;
            if ($deferred) {
                unset($this->futureMap[$wsMessage->id]);
                Logger::debug("Unified WebSocket API response received, id={$wsMessage->id}, op={$wsMessage->op}");
                try {
                    $deferred->resolve($wsMessage);
                } catch (Exception $e) {
                    Logger::error("Unified WebSocket API response processing failed, id={$wsMessage->id}, op={$wsMessage->op}", ['exception' => $e]);
                }
            }
        });

        $this->client->on("reconnected", function () {
            // No recovery needed for unified API; subscriptions are managed elsewhere
        });
    }

    /**
     * {@inheritdoc}
     */
    public function start(): PromiseInterface
    {
        return $this->client->start();
    }

    /**
     * {@inheritdoc}
     */
    public function stop(): PromiseInterface
    {
        return $this->client->stop();
    }

    /**
     * {@inheritdoc}
     */
    public function call(UnifiedWsArgs $args): PromiseInterface
    {
        // Ensure ID exists
        if (empty($args->id)) {
            $args->id = uniqid('', true);
        }

        // Ensure operation type is set
        if (empty($args->op)) {
            throw new RuntimeException("Missing required field: op");
        }

        $deferred = new Deferred();

        // Check for duplicate ID
        if (isset($this->futureMap[$args->id])) {
            throw new RuntimeException("Duplicate request id: " . $args->id);
        }

        $this->futureMap[$args->id] = $deferred;

        Logger::info("Calling unified WebSocket API, id={$args->id}, op={$args->op}");

        $this->client->write($args->id, $args, $this->option->writeTimeout)
            ->then(
                function () use ($args, $deferred) {
                    Logger::info("Unified WebSocket API call sent, id={$args->id}, op={$args->op}");
                },
                function ($ex) use ($args, $deferred) {
                    unset($this->futureMap[$args->id]);
                    Logger::error("Unified WebSocket API call failed, id={$args->id}, op={$args->op}", ['exception' => $ex]);
                    $deferred->reject($ex);
                }
            );

        return $deferred->promise();
    }

    /**
     * Notify event listeners of WebSocket events
     *
     * @param string $ev
     * @param string $msg
     */
    private function notifyEvent(string $ev, string $msg): void
    {
        if (isset($this->option->eventCallback)) {
            try {
                call_user_func($this->option->eventCallback, $ev, $msg);
            } catch (Exception $e) {
                Logger::error("Exception when notifying event", ['exception' => $e]);
            }
        }
    }
}